### PR TITLE
[Feature](Job)Provide unified internal Job scheduling

### PR DESCRIPTION
**Pull request by CalvinKirs**
Support internal event registration, and timing scheduling

## Proposed changes
Doris Event Scheduler is a task that runs according to a plan. It provides a function similar to Cron in Unix, that is, to execute a command at a certain time in the future.
For database-type software, they usually have similar functions. For example, MySQL also provides the Event Scheduler function, while AWS RedShift uses the unified scheduled task function to complete similar functions.

The Event Scheduler function is as its name suggests. In Doris, we have the following scenarios that can be used.
like:

- Periodic Insert into
- Asynchronous insert into
- Refresh the Catalog periodically
- Micro-batch broker load
- Timing & periodic refresh of materialized views

image

We use the time wheel algorithm to complete the scheduling and triggering of periodic tasks. The implementation of the time wheel algorithm refers to netty's HashedWheelTimer.
We will periodically (10 minutes by default) put the events that need to be triggered in the future cycle into the time wheel for periodic scheduling. In order to ensure the efficient triggering of tasks and avoid task blocking and subsequent task scheduling delays, we use Disruptor to implement the production and consumption model.
When the task expires and needs to be triggered, the task will be put into the RingBuffer of the Disruptor, and then the consumer thread will consume the task.

Consumers need to register for events, and event registration needs to provide event executors. Event executors are a functional interface with only one method for executing events.

If it is a single event, the event definition will be deleted after the scheduling is completed; if it is a periodic event, it will be put back into the time wheel according to the periodic scheduling after the scheduling is completed.
- User registration event
- The event scheduler puts events into the time wheel for periodic scheduling
- After the event expires, put the event into the Disruptor's RingBuffer
- The consumer thread consumes events from the RingBuffer
```java
/**
 * This interface provides a contract for registering timed scheduling events.
 * The implementation should trigger events in a timely manner using a specific algorithm.
 * The execution of the events may be asynchronous and not guarantee strict timing accuracy.
 */
public interface EventJobSchedulerRegister {

    /**
     * Register a event job
     *
     * @param name       event job name,it's not unique
     * @param intervalMs event job interval, unit: ms
     * @param executor   event job executor @See {@link EventJobExecutor}
     * @return event job id
     */
    Long registerEventJob(String name, Long intervalMs, EventJobExecutor executor);

    /**
     * Register a event job
     *
     * @param name           event job name,it's not unique
     * @param intervalMs     event job interval, unit: ms
     * @param startTimeStamp event job start time stamp, unit: ms
     *                       if startTimeStamp is null, event job will start immediately in the next cycle
     *                       startTimeStamp should be greater than current time
     * @param executor       event job executor @See {@link EventJobExecutor}
     * @return event job id
     */
    Long registerEventJob(String name, Long intervalMs, Long startTimeStamp, EventJobExecutor executor);


    /**
     * Register a event job
     *
     * @param name           event job name,it's not unique
     * @param intervalMs     event job interval, unit: ms
     * @param startTimeStamp event job start time stamp, unit: ms
     *                       if startTimeStamp is null, event job will start immediately in the next cycle
     *                       startTimeStamp should be greater than current time
     * @param endTimeStamp   event job end time stamp, unit: ms
     *                       if endTimeStamp is null, event job will never stop
     *                       endTimeStamp must be greater than startTimeStamp and endTimeStamp should be greater
     *                       than current time
     * @param executor       event job executor @See {@link EventJobExecutor}
     * @return event job id
     */
    Long registerEventJob(String name, Long intervalMs, Long startTimeStamp, Long endTimeStamp,
                          EventJobExecutor executor);

    /**
     * if event job is running, pause it
     * pause means event job will not be executed in the next cycle,but current cycle will not be interrupted
     * we can resume it by {@link #resumeEventJob(Long)}
     *
     * @param eventId event job id
     *                if eventId not exist, return false
     * @return true if pause success, false if pause failed
     */
    Boolean pauseEventJob(Long eventId);

    /**
     * if event job is running, stop it
     * stop means event job will not be executed in the next cycle and current cycle will be interrupted
     * stop not can be resumed, if you want to resume it, you should register it again
     * we will delete stopped event job
     *
     * @param eventId event job id
     * @return true if stop success, false if stop failed
     */
    Boolean stopEventJob(Long eventId);

    /**
     * if event job is paused, resume it
     *
     * @param eventId event job id
     * @return true if resume success, false if resume failed
     */
    Boolean resumeEventJob(Long eventId);

}
```
#### What can we do with it?
- Provide user-oriented creation of timed/one-time jobs
```
CREATE
    [DEFINER = user]
    EVENT
    [IF NOT EXISTS]
    event_name
    ON SCHEDULE schedule
    [ON COMPLETION [NOT] PRESERVE]
    [ENABLE | DISABLE | DISABLE ON SLAVE]
    [COMMENT 'string']
    DO event_body;

schedule: {
    AT timestamp [+ INTERVAL interval] ...
  | EVERY interval
    [STARTS timestamp [+ INTERVAL interval] ...]
    [ENDS timestamp [+ INTERVAL interval] ...]
}

interval:
    quantity { WEEK | DAY | HOUR | MINUTE |
              | SECOND }
```
eg. You want to use the method of catalog + insert into to import the full amount of data
```
CREATE EVENT load_order
ON SCHEDULER 
AT '2023-05-10 03:00:00'
DO
 INSERT INTO user_activity FROM SELECT * FROM activity.user.activity /
```
- Timed refresh of materialized views
This requires internal adaptation of the materialized view with this function
image

### CODE REVIEW

**Comment by CalvinKirs** [comment]
Todo UT

**Comment by yiguolei** [comment]
Could you please provide some SQL usage docs?

**Comment by CalvinKirs** [comment]
> Could you please provide some SQL usage docs?↳

At present, it only provides internal timing scheduling, and the registration of events through sql on the client side has not yet been developed.

**Comment by CalvinKirs** [comment]
run buildall

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 44.66 seconds
 stream load tsv:          461 seconds loaded 74807831229 Bytes, about 154 MB/s
 stream load json:         21 seconds loaded 2358488459 Bytes, about 107 MB/s
 stream load orc:          57 seconds loaded 1101869774 Bytes, about 18 MB/s
 stream load parquet:          28 seconds loaded 861443392 Bytes, about 29 MB/s
 insert into select:          69.0 seconds inserted 10000000 Rows, about 144K ops/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230625095113_clickbench_pr_167092.html

**Comment by CalvinKirs** [comment]
run P1 Regression

**Comment by CalvinKirs** [comment]
run buildall

**Comment by CalvinKirs** [comment]
run buildall

**Review by dataroaring — APPROVED** [review]
LGTM

**Comment by CalvinKirs** [comment]
run P0 Regression

**Comment by CalvinKirs** [comment]
run clickbench (benchmark)

**Comment by CalvinKirs** [comment]
run buildall

**Review by zy-kkk — APPROVED** [review]

File: fe/fe-core/src/main/java/org/apache/doris/scheduler/AsyncJobRegister.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+public class AsyncJobRegister implements JobRegister {
+
+    private final AsyncJobManager asyncJobManager;
+
+    public AsyncJobRegister() {
+        this.asyncJobManager = new AsyncJobManager();
+    }
+
+    @Override
+    public Long registerJob(String name, Long intervalMs, JobExecutor executor) {
```
[the remark this answers is not in the document]
Comment (CalvinKirs) [reply fe/fe-core/src/main/java/org/apache/doris/scheduler/AsyncJobRegister.java:46]: Yes, this requires providing additional methods. I'll add this when we do the materialized view

File: fe/fe-core/src/main/java/org/apache/doris/scheduler/AsyncJobRegister.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+        return this.registerJob(name, intervalMs, null, null, executor);
+    }
+
+    @Override
+    public Long registerJob(String name, Long intervalMs, Long startTimeStamp, JobExecutor executor) {
+        return this.registerJob(name, intervalMs, startTimeStamp, null, executor);
+    }
+
+    @Override
+    public Long registerJob(String name, Long intervalMs, Long startTimeStamp, Long endTimeStamp,
```
[the remark this answers is not in the document]
Comment (CalvinKirs) [reply fe/fe-core/src/main/java/org/apache/doris/scheduler/AsyncJobRegister.java:56]: Yes, in the future we need to distinguish between job tasks directly submitted by users and system jobs and internal jobs.

File: fe/fe-core/src/main/java/org/apache/doris/scheduler/constants/JobStatus.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+package org.apache.doris.scheduler.constants;
+
+public enum JobStatus {
+
+    /**
+     * When the task is not started, the initial state will be triggered.
+     * The initial state can be started
+     */
+    RUNNING,
```
[the remark this answers is not in the document]
Comment (CalvinKirs) [reply fe/fe-core/src/main/java/org/apache/doris/scheduler/constants/JobStatus.java:26]: good idea
currently goes to the stop state when the job is complete.
I would add a completed state

**Review by zddr — APPROVED** [review]
Further improvement of other PR

**Review by CalvinKirs — COMMENTED** [review]
trigger

**Pull request merged**